File: backend/app/services/auth_service.py
"""Authentication service — user registration, login, token generation."""

from __future__ import annotations

import secrets
from uuid import UUID

import httpx
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.security import create_access_token, hash_password, verify_password
from app.models.user import User


class AuthService:
    """Stateless helper methods for auth workflows."""

    @staticmethod
    async def register(db: AsyncSession, *, email: str, password: str, name: str) -> User:
        user = User(
            email=email,
            password_hash=hash_password(password),
            name=name,
        )
        db.add(user)
        await db.flush()
        await db.refresh(user)
        return user

    @staticmethod
    async def authenticate_user(db: AsyncSession, email: str, password: str) -> str | None:
        """Return a JWT access token if credentials are valid, else None."""
        result = await db.execute(select(User).where(User.email == email))
        user = result.scalar_one_or_none()
        if user is None or not verify_password(password, user.password_hash):
            return None
        return create_access_token(data={"sub": str(user.id)})

    @staticmethod
    async def authenticate_social(db: AsyncSession, supabase_access_token: str) -> tuple[str, bool] | None:
        """Verify a Supabase session token (issued client-side after a
        Google/Facebook OAuth login via supabase_flutter), then find or
        create a matching local User and issue our own JWT for it.

        Returns (access_token, is_new_user) or None if the Supabase token
        couldn't be verified.
        """
        if not settings.SUPABASE_URL or not settings.SUPABASE_ANON_KEY:
            # Social login isn't configured on this deployment yet.
            return None

        async with httpx.AsyncClient(timeout=10.0) as client:
            try:
                response = await client.get(
                    f"{settings.SUPABASE_URL}/auth/v1/user",
                    headers={
                        "Authorization": f"Bearer {supabase_access_token}",
                        "apikey": settings.SUPABASE_ANON_KEY,
                    },
                )
            except httpx.HTTPError:
                return None

        if response.status_code != 200:
            return None

        payload = response.json()
        email = payload.get("email")
        if not email:
            return None

        metadata = payload.get("user_metadata") or {}
        name = (
            metadata.get("full_name")
            or metadata.get("name")
            or email.split("@")[0]
        )

        result = await db.execute(select(User).where(User.email == email))
        user = result.scalar_one_or_none()
        is_new_user = False

        if user is None:
            # Social-only accounts get a random, never-shared password
            # hash so the column stays NOT NULL; they simply never log
            # in with a password unless they later set one.
            user = User(
                email=email,
                password_hash=hash_password(secrets.token_urlsafe(32)),
                name=name,
            )
            db.add(user)
            await db.flush()
            await db.refresh(user)
            is_new_user = True

        token = create_access_token(data={"sub": str(user.id)})
        return token, is_new_user

    @staticmethod
    async def get_user_by_id(db: AsyncSession, user_id: UUID) -> User | None:
        result = await db.execute(select(User).where(User.id == user_id))
        return result.scalar_one_or_none()

File: mobile/lib/core/api/auth_service.dart
import 'package:dio/dio.dart';
import 'api_client.dart';

class AuthUser {
  final String id;
  final String email;
  final String name;
  const AuthUser({required this.id, required this.email, required this.name});

  factory AuthUser.fromJson(Map<String, dynamic> json) {
    return AuthUser(
      id: json['id'] as String,
      email: json['email'] as String,
      name: json['name'] as String,
    );
  }
}

/// Result of exchanging a Supabase OAuth session for our own backend
/// token. [isNewUser] tells the caller whether to route through
/// onboarding (first-ever sign-in) or straight to home.
class SocialLoginResult {
  final String token;
  final bool isNewUser;
  const SocialLoginResult({required this.token, required this.isNewUser});
}

/// Talks to the backend's /auth/register, /auth/login and /auth/social
/// endpoints.
class AuthService {
  AuthService(this._api);

  final ApiClient _api;

  /// Registers a new account. Throws [ApiException]/[ServerException] on failure.
  Future<AuthUser> register({
    required String name,
    required String email,
    required String password,
  }) async {
    final response = await _api.post<Map<String, dynamic>>(
      '/auth/register',
      data: {
        'name': name,
        'email': email,
        'password': password,
      },
    );
    return AuthUser.fromJson(response.data!);
  }

  /// Logs in with email/password. The backend expects OAuth2 form-encoded
  /// data with the email passed as `username`.
  ///
  /// Returns the bearer token on success.
  Future<String> login({
    required String email,
    required String password,
  }) async {
    final response = await _api.post<Map<String, dynamic>>(
      '/auth/login',
      data: {
        'username': email,
        'password': password,
      },
      options: Options(
        contentType: Headers.formUrlEncodedContentType,
      ),
    );
    final token = response.data!['access_token'] as String;
    return token;
  }

  /// Exchanges a Supabase session access token (obtained client-side
  /// after a Google/Facebook OAuth login via supabase_flutter) for our
  /// own backend bearer token. The backend verifies the token against
  /// Supabase and finds-or-creates the matching local user.
  Future<SocialLoginResult> loginWithSupabaseToken(String supabaseAccessToken) async {
    final response = await _api.post<Map<String, dynamic>>(
      '/auth/social',
      data: {
        'access_token': supabaseAccessToken,
      },
    );
    final data = response.data!;
    return SocialLoginResult(
      token: data['access_token'] as String,
      isNewUser: data['is_new_user'] as bool? ?? false,
    );
  }
}
